test/onnx/gen_onnx.py

```diff
@@ -9695,6 +9695,663 @@ def softmax_dyn_test():
+            "1",
+            "2",
+        ],
+        outputs=["3"],
+        reduction="mean",
+    )
+
+    return ([node], [scores, labels, weights], [loss])
+
+
+@onnx_test()
+def softmaxcrossentropyloss_2d_mean_reduction_double_weighted_test():
+    scores = helper.make_tensor_value_info('0', TensorProto.DOUBLE, [4, 4])
+    labels = helper.make_tensor_value_info('1', TensorProto.INT32, [4])
+    weights = helper.make_tensor_value_info('2', TensorProto.DOUBLE, [4])
+    loss = helper.make_tensor_value_info('3', TensorProto.DOUBLE, [4])
+
+    node = onnx.helper.make_node(
+        "SoftmaxCrossEntropyLoss",
+        inputs=[
+            "0",
+            "1",
+            "2",
+        ],
+        outputs=["3"],
+        reduction="mean",
+    )
+
+    return ([node], [scores, labels, weights], [loss])
+
+
+@onnx_test()
+def softmaxcrossentropyloss_2d_mean_reduction_half_weighted_test():
+    scores = helper.make_tensor_value_info('0', TensorProto.FLOAT16, [4, 4])
+    labels = helper.make_tensor_value_info('1', TensorProto.INT32, [4])
+    weights = helper.make_tensor_value_info('2', TensorProto.FLOAT16, [4])
+    loss = helper.make_tensor_value_info('3', TensorProto.FLOAT16, [4])
+
+    node = onnx.helper.make_node(
+        "SoftmaxCrossEntropyLoss",
+        inputs=[
+            "0",
+            "1",
+            "2",
+        ],
+        outputs=["3"],
+        reduction="mean",
+    )
+
+    return ([node], [scores, labels, weights], [loss])
+
+
+@onnx_test()
+def softmaxcrossentropyloss_kd_mean_reduction_half_weighted_test():
+    scores = helper.make_tensor_value_info('0', TensorProto.FLOAT16,
+                                           [4, 4, 2, 2])
+    labels = helper.make_tensor_value_info('1', TensorProto.INT32, [4, 2, 2])
+    weights = helper.make_tensor_value_info('2', TensorProto.FLOAT16, [4])
+    loss = helper.make_tensor_value_info('3', TensorProto.FLOAT16, [4])
+
+    node = onnx.helper.make_node(
+        "SoftmaxCrossEntropyLoss",
+        inputs=[
+            "0",
+            "1",
+            "2",
+        ],
+        outputs=["3"],
+        reduction="mean",
+    )
+
+    return ([node], [scores, labels, weights], [loss])
+
+
+@onnx_test()
+def softmaxcrossentropyloss_kd_sum_reduction_double_weighted_test():
+    scores = helper.make_tensor_value_info('0', TensorProto.DOUBLE,
+                                           [4, 4, 2, 2])
+    labels = helper.make_tensor_value_info('1', TensorProto.INT32, [4, 2, 2])
+    weights = helper.make_tensor_value_info('2', TensorProto.DOUBLE, [4])
+    loss = helper.make_tensor_value_info('3', TensorProto.DOUBLE, [4])
+
+    node = onnx.helper.make_node(
+        "SoftmaxCrossEntropyLoss",
+        inputs=[
+            "0",
+            "1",
+            "2",
+        ],
+        outputs=["3"],
+        reduction="sum",
+    )
+
+    return ([node], [scores, labels, weights], [loss])
+
+
+@onnx_test()
+def softmaxcrossentropyloss_kd_no_reduction_weighted_test():
```

Comment:
can you add tests

1.  where ignore_index is negative 
2. ignore index is out of bound 
3. matrix is 4D

Reply:
Yes I can do that

Reply:
Added three cases for the ignore out of bound  (positive and negative bounds) as well as the negative index.

Will add 4d matrix later this evening.

Reply:
Added 4d tests. Had these before but hitting issue with verify when I compare this to ref data. Working on solving these and the asymmetric input cases that I seem to be failing before I'll go back and fix the parser tests. Parser tests sill need an update as now ignore_index literal isn't added when it's not detected on input.

```
[==========] 25 tests ran
[  FAILED  ] 9 tests failed
[  FAILED  ] softmaxcrossentropyloss_kd_no_reduction_even_weighted_ones_test
[  FAILED  ] softmaxcrossentropyloss_kd_no_reduction_uneven_weighted_ones_test
[  FAILED  ] softmaxcrossentropyloss_kd_sum_reduction_weighted_test_ones
[  FAILED  ] softmaxcrossentropyloss_kd_sum_reduction_weighted_test_zeroes
[  FAILED  ] softmaxcrossentropyloss_kd_mean_reduction_weighted_test
[  FAILED  ] softmaxcrossentropyloss_kd_mean_reduction_uneven_weighted_test
[  FAILED  ] softmaxcrossentropyloss_2d_no_reduction_test_asym_test
[  FAILED  ] softmaxcrossentropyloss_2d_no_reduction_uneven_weighted_ones_asym_batch
[  FAILED  ] softmaxcrossentropyloss_2d_no_reduction_uneven_weighted_ones_asym_class
```

Reply:
Please note that an `ignore_index` that is negative is actually the same case as out_of_bounds.

Reply:
Solved